modules/javafx.controls/src/main/java/com/sun/javafx/scene/control/ResizeHelper.java

```diff
@@ -67,81 +69,191 @@ public ResizeHelper(ResizeFeaturesBase rf,
 
         for (int i = 0; i < count; i++) {
             TableColumnBase<?,?> c = columns.get(i);
-            size[i] = c.getWidth();
+            size[i] = snapCeil(c.getWidth());
 
             if (c.isResizable()) {
-                double cmin = c.getMinWidth();
-                double cmax = c.getMaxWidth();
+                double cmin = snapCeil(c.getMinWidth()); // always honor min width!
+                double cmax = snapFloor(c.getMaxWidth()); // TableColumnBase.doSetWidth() clamps to (min,max) range
                 min[i] = cmin;
                 max[i] = cmax;
-                pref[i] = clip(c.getPrefWidth(), cmin, cmax);
+                pref[i] = clip(snapRound(c.getPrefWidth()), cmin, cmax);
+                // skip fixed columns
+                if (cmin == cmax) {
+                    skip.set(i, true);
+                }
             } else {
                 skip.set(i, true);
             }
         }
     }
 
     public void resizeToContentWidth() {
-        boolean needsAnotherPass;
-        do {
-            needsAnotherPass = false;
-            double sumWidths = 0.0;
-            double sumMins = 0;
-            for (int i = 0; i < count; i++) {
-                sumWidths += size[i];
-                sumMins += min[i];
+        if (skip.cardinality() == count) {
+            return;
+        }
+
+        // compute delta (snapped)
+        // if delta < 0 (and sum(width) < sum(pref)) -> distribute from size to min
+        // if delta < 0 (and sum(width) > sum(pref)) -> distribute from size to pref
+        // if delta > 0 (and sum(width) < sum(pref)) -> distribute from size to pref
+        // else -> distribute from size to max
+        //
+        double sumWidths = sum(size);
+        double sumPrefs = sum(pref);
+
+        double delta = target - sumWidths;
+        if (delta < 0.0) {
+            // shrink
+            if (target < sumPrefs) {
+                distribute(delta, min);
+            } else {
+                distribute(delta, pref);
+            }
+        } else if (delta > 0.0) {
```

Comment:
Do we need delta = 0.0 condition here?

Reply:
excellent question.  we don't - zero delta means no change, so no action is needed.